jetbrains-core/src/software/aws/toolkits/jetbrains/core/tools/ToolType.kt

```diff
@@ -0,0 +1,46 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.core.tools
+
+import java.nio.file.Path
+
+/**
+ * Represents an executable external tool such as a CLI
+ *
+ * Note: It is recommended that all implementations of this interface are stateless and are an `object`
+ */
+interface ToolType<VersionScheme : Version> {
+    /**
+     * ID used to represent the executable in caches and settings. Must be globally unique
+     */
+    val id: String
+
+    /**
+     * Name of the executable for users, e.g. the marketing name of the executable
+     */
+    val displayName: String
+
+    /**
+     * List of supported [VersionRange]. An empty list means any version is supported
+     */
+    fun supportedVersions(): List<VersionRange<VersionScheme>> = emptyList()
```

Comment:
I saw your "ahem" in the PR description - by why do we need a list of ranges?

The CLI example - is that we have a supported range for CLI v1 and CLI v2 ?

Reply:
Yes, the AWS CLI is the only use case today, where ECS Exec has a min in the V1 branch and and a min in the v2 branch

Reply:
We could use stricterMinVersion to gate it though. Thoughts?

```
val newMin = if(version.majorVersion == 1) SemanticVersion(1, ..) else SemanticaVersion(2...)
```

Reply:
Removed multiple ranges support